### PR TITLE
Purging missing classes array on autoload change

### PR DESCRIPTION
**Pull request by moufmouf**
PR #5558 added a "missing classes" array in the autoloader.

This array caches missing classes for optimization purpose.

However, this array assumes that methods that affect the classmap (like `addClassMap`, `addPsr4`...) are performed **before** the autoloader is actually used.

This is almost always true.... except for a few edge cases like Drupal 8. Haaaa Drupal :(

Drupal has a powerful module system where modules can be enabled/disabled by the user. When you enable a module, Drupal dynamically adds the PSR4 autoloader for the module into the Drupal autoloader (even if the composer autoloader has already been used before).

In some cases, Drupal performed a "class_exists" call on a given class (say class `Foo`). It does not exists. A few functions later, it performs a call to `addPsr4` (because it is in the process of enabling a module). Now, if the newly enabled module tries to access the `Foo` class, Composer autoloader will not autoload it (because it is still cached in the `$missingClasses` cache.

This PR simply resets the `$missingClasses` cache every time the autoloader is modified (because the classes in the array might now be accessible).

It also comes with the regular unit test that would fail if this PR is not applied.

### CODE REVIEW

**Comment by stof** [comment]
you should use a separate autoloader instance for that, not the main composer one though, because altering its config after loading ``vendor/autoload.php`` would kill the optimizations being applied.

**Comment by moufmouf** [comment]
Hey @stof,

Are you sure? For the 99.99% of the persons that use autoloader "normally", do you think this PR kills the optimizations being applied? (my understanding was that as long as I'm not touching the `$classMap`, I'm not creating a COW so I'm ok). For the very very vast majority of users, `$missingClasses` is empty when calls to `addPsr4` et al. is called.

For the 0.01% remaining, those optimizations should be disabled because they are bogus (it is what this PR does).

Also, I don't really have the power to ask the Drupal project to not use Composer autoloader (and I think it would be a mistake for them :) )

**Comment by stof** [comment]
@moufmouf what kills the optimization (not the one about cache misses but previous ones) is changing the composer autoloader config (adding new paths, etc...) as it forces PHP to copy the array out of shared memory to be able to alter it.
the autoloader instance returned by ``vendor/autoload.php`` should be considered immutable if you care about performance.

If you have a need for dynamic autoloader configuration, you should use a second autoloader for this part (PHP is perfectly fine with having several autoloaders registered)

**Comment by moufmouf** [comment]
tl;dr:

I understand that people should not do this. The fact is they do (it's not my code, it's Drupal's :) )
I don't believe this PR is impacting performance for existing users, except the ones that using Composer as a dynamic autoloader, but this is actually ok because we are protecting those from a bug.

---

> the autoloader instance returned by vendor/autoload.php should be considered immutable if you care about performance.

Looking at Drupal 8 code, this call to `addPsr4` performed after autoloading some classes is performed only when "enabling" a module. So they clearly do not care about the few ms lost here (this happens only when deploying a new module)

> If you have a need for dynamic autoloader configuration, you should use a second autoloader for this part (PHP is perfectly fine with having several autoloaders registered)

I know PHP can have several autoloaders. And I still don't think it is worthwhile to use a different autoloader than composer simply for this use case.

I understand what you say about considering the autoloader as immutable but this is not stated anywhere and apparently, Drupal devs still do it, leaving me with this issue. If the autoloader is really to be considered as immutable, Composer could enforce a "frozen" state of the autoloader. Or maybe splitting the autoloader in a "autoloader builder" and an "autoloader" just like Symfony does with the its container would be an option?

**Comment by stof** [comment]
@moufmouf you can use a separate instance of the Composer ClassLoader.

However, this PR indeed makes sense for people using their own mutable autoloader. I would reject a similar PR resetting the APCu cache though (if people use the new ``--apcu-autoloader`` flag, they should really avoid altering the ClassLoader config for the dumped instance)

**Comment by moufmouf** [comment]
> you can use a separate instance of the Composer ClassLoader.

@stof Thanks for the tip, you put me on the right track. This is not completely straightforward because the composer autoloader creation code is hidden in an autogenerated singleton... like this:

```php
class ComposerAutoloaderInit452400d9c73e77a27f816d92b69d76c4
{
    private static $loader;
//...
    public static function getLoader()
    {
        if (null !== self::$loader) {
            return self::$loader;
        }
//...
```

I found a workaround though.... Just before performing my call to `class_exists`, I'm cloning the existing Composer autoloader and prepending the clone to the autoloader queue. That way, the clone `$missingClasses` gets filled. Just after the `class_exists` call, I'm unregistering the cloned autoloader. Tada! (probably one of the worst hacks of my life :) ) 

So yeah... I get this workaround but I still think this PR could be useful to some people.

> I would reject a similar PR resetting the APCu cache though (if people use the new --apcu-autoloader flag, they should really avoid altering the ClassLoader config for the dumped instance)

:+1:

**Comment by stof** [comment]
@moufmouf creating a new instance is just ``new Composer\ClassLoader``. You should not clone the existing configuration in the new instance. You should configure only the **additional** dynamic paths in your new instance.

**Comment by Seldaek** [comment]
What @stof says, basically all you need is something like this:

```
$moduleLoader = new Composer\Autoload\ClassLoader;
$moduleLoader->addPsr4(...); // only the new module config
$moduleLoader->register();
```

After this, the module's classes are autoloadable.
